drivers/sensor/bma4xx/bma4xx_emul.c

```diff
@@ -0,0 +1,428 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "bma4xx_regs.h"
+#include "bma4xx_emul.h"
+#include "zephyr/toolchain.h"
+
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c_emul.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/math_extras.h>
+
+#define DT_DRV_COMPAT bosch_bma4xx
+
+LOG_MODULE_REGISTER(bosch_bma4xx, LOG_LEVEL_INF);
+
+/* The total number of used registers specified in bma4xx datasheet. */
+#define NUM_REGS 0x7E
+
+struct bma4xx_emul_data {
+	/* Holds register data. */
+	uint8_t regs[NUM_REGS];
+	/** True if the sensor is currently enabled. */
+	bool accel_enabled;
+	/** Current sensor range, ±2/4/8/16g; value is positive gs. */
+	uint8_t accel_range;
+	/** Raw register value of ACC_CONF:acc_odr. */
+	uint8_t odr_raw;
+	/** Current sensor reading on XYZ axes, in milli-g. */
+	int acceleration[3];
+	/** Axis offset register values, XYZ. */
+	uint8_t offset[3];
+	/** NV_CONF register value. */
+	uint8_t nv_config;
+	/**
+	 * True if the sensor FIFO is currently enabled.
+	 *
+	 * Only headerless mode for the accelerometer alone is supported.
+	 */
+	bool fifo_enabled;
```

Comment:
A lot of these are duplicates, the values exist inside the `regs[]` array. Instead we should have getters to check the value otherwise it could get difficult to manage the 2 states.